pkg/providers/ibmcloud.go

```diff
@@ -0,0 +1,331 @@
+
+	// TODO: Would love to find a way to list locations where the package is available but
+	// I can't for the life of me figure it out, need to revisit
+	// I've removed some of the more expensive locations from the list below
+	locations := []string{"DALLAS09", "DALLAS10", "DALLAS12", "DALLAS13", "FRANKFURT", "FRANKFURT04", "FRANKFURT05", "LONDON02", "LONDON04",
+		"LONDON05", "LONDON06", "MILAN", "MONTREAL", "PARIS", "MADRID02", "MADRID04", "MADRID05", "SANJOSE03", "SANJOSE04", "TORONTO",
+		"TORONTO04", "TORONTO05", "WASHINGTON06", "WASHINGTON07"}
+	rand.Shuffle(len(locations), func(i, j int) {
+		locations[i], locations[j] = locations[j], locations[i]
+	})
+
+	getLocation := func() (string, error) {
+		if len(locations) == 0 {
+			return "", errors.New("something went wrong")
+		}
+		return locations[0], nil
+	}
+	location, err := getLocation()
+
+	// Package 200 are the fast provisioning servers (40-60min provsion)
+	packageId := 200
+	hostname := strings.Split(uuid.New().String(), "-")[0]
+	domain := pool + ".ofcir"
+
+	key, err := p.ensureSSHKey(p.config.Sshkey)
+	if err != nil {
+		return "", err
+	}
+
+	server := []datatypes.Hardware{
+		{
+			Hostname: sl.String(hostname),
+			Domain:   sl.String(domain),
+		},
+	}
+
+	// Get a list of available keys with
+	// ibmcloud sl order item-list  BARE_METAL_SERVER
+	required_items := []string{
+		"REBOOT_KVM_OVER_IP",
+		"UNLIMITED_SSL_VPN_USERS_1_PPTP_VPN_USER_PER_ACCOUNT",
+		"OS_CENTOS_STREAM_8_X_64_BIT",
+		"BANDWIDTH_0_GB_2",
+		"1_GBPS_PUBLIC_PRIVATE_NETWORK_UPLINKS",
+		"1_IP_ADDRESS",
+	}
+	// Build a skeleton SoftLayer_Product_Item_Price objects.
+	prices := p.getItemPriceList(packageId, required_items)
+
+	// We need to find the ID of the preset with the required name
+	packageService := services.GetProductPackageService(p.client)
+
+	// ibmcloud sl order preset-list  --prices BARE_METAL_SERVER
+	// e.g. presetname == 1U_4210S_384GB_2X4TB_RAID_1
+	// 1U_4210S_384GB_2X4TB_RAID_1, id = 1278
+	presets, err := packageService.Id(packageId).GetActivePresets()
+	presetId := -1
+	for _, preset := range presets {
+		if *preset.KeyName == presetname {
+			presetId = *preset.Id
+			break
+		}
+	}
+	if presetId == -1 {
+		return "", errors.New("Can't find preset with name " + presetname)
+	}
+
+	orderkeys := []datatypes.Container_Product_Order_SshKeys{
+		{
+			SshKeyIds: []int{*key},
+		},
+	}
+
+	// Build a container_Product_Order object.
+	orderTemplate := datatypes.Container_Product_Order{
+		Quantity:         sl.Int(1),
+		Location:         sl.String(location),
+		PackageId:        sl.Int(packageId),
+		PresetId:         sl.Int(presetId),
+		Prices:           prices,
+		Hardware:         server,
+		UseHourlyPricing: sl.Bool(true),
+		SshKeys:          orderkeys,
+	}
+
+	// Get SoftLayer_Product_Order service.
+	service := services.GetProductOrderService(p.client)
+
+	//result, err := service.VerifyOrder(&orderTemplate)
```

Comment:
typo?

Reply:
I left this here on purpose so I don't forget how to do a order "dry run" while testing, will add a comment